bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java

```diff
@@ -37,13 +54,49 @@
     private RuleEngine ruleEngine;
     private Logger logger;
     private Storage<Boolean> disabledRulesStorage;
+    private TemplateManager templateManager;
+
+    /**
+     * {@link Map} of template UIDs to rules where these templates participated.
+     */
+    private Map<String, Set<String>> mapTemplateToRules = new HashMap<String, Set<String>>();
+    @SuppressWarnings("rawtypes")
+    private ServiceTracker templateProviderTracker;
 
     private static final String SOURCE = RuleRegistryImpl.class.getSimpleName();
 
-    public RuleRegistryImpl(RuleEngine ruleEngine) {
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    public RuleRegistryImpl(RuleEngine ruleEngine, TemplateManager tManager, final BundleContext bc) {
         logger = LoggerFactory.getLogger(getClass());
         this.ruleEngine = ruleEngine;
+        this.templateManager = tManager;
         ruleEngine.setStatusInfoCallback(this);
+        templateProviderTracker = new ServiceTracker(bc, TemplateProvider.class.getName(),
+                new ServiceTrackerCustomizer() {
+
+                    @Override
+                    public Object addingService(ServiceReference reference) {
+                        TemplateProvider provider = bc.getService(reference);
+                        templateUpdated(provider.getTemplates(null));
+                        return provider;
+                    }
+
+                    @Override
+                    public void modifiedService(ServiceReference reference, Object service) {
+                        TemplateProvider provider = bc.getService(reference);
+                        templateUpdated(provider.getTemplates(null));
+
+                    }
+
+                    @Override
+                    public void removedService(ServiceReference reference, Object service) {
+                        // TODO Auto-generated method stub
```

Comment:
Remove is not used, because this tracker only notifies RuleRegistry about new available templates and RuleRegistry has to try to resolve some of not resolved rules created by templates. When template provider disappears it has not any effect of resolved rules and for unresolved  (waiting templates) ones.